http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala

```diff
@@ -164,7 +165,10 @@ final class LeoAppServiceInterp[F[_]: Parallel](config: AppServiceConfig,
         getSavableCluster(originatingUserEmail, cloudContext, ctx.now)
       )
 
+      _ <- clusterSemaphore.acquire
```

Comment:
I'm not sure this is a good pattern -- if `KubernetesServiceDbQueries.saveOrGetClusterForApp(saveCluster).transaction` fails for whatever reason it may never release the lock. I think a better pattern is `clusterSemaphore.permit.use(…)` which wraps the acquire/release in a cats-effect `Resource`.

Also you might look into using [KeyLock](https://github.com/DataBiosphere/leonardo/blob/develop/core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/KeyLock.scala) which already exists in Leo -- it basically does the same thing, but is able to lock on a per-key basis (since we probably don't need a global lock, but a lock per-cloud-context).

Last comment: maybe it's worth trying a `Serializable` transaction to see if it resolves the issue, and not using a semaphore at all? It should be as easy as:
```
saveClusterResult <- KubernetesServiceDbQueries.saveOrGetClusterForApp(saveCluster).transaction(isolationLevel = TransactionIsolation.Serializable)
```

Reply:
Ah, this is what I was probably looking for. For some reason, I thought anything in the context of a DBIO in a for comprehension (aka a flatmapped DBIO) was atomic. I knew there had be a way to attempt to enforce a DBIO to be atomic